fabric-data-generation-api-v1/src/client/java/net/fabricmc/fabric/impl/datagen/client/SoundTypeBuilderImpl.java

```diff
@@ -0,0 +1,217 @@
+
+		for (Entry sound : sounds) {
+			if (sound.type() == RegistrationType.SOUND_EVENT) {
+				Registries.SOUND_EVENT.getOptionalValue(sound.name()).orElseThrow(() -> new IllegalStateException("Referenced sound event " + sound.name() + " does not exist"));
+			}
+		}
+
+		return new SoundType(sounds, category, Optional.ofNullable(subtitle));
+	}
+
+	public record SoundType(List<Entry> sounds, SoundCategory category, Optional<String> subtitle) {
+		private static final Map<String, SoundCategory> CATEGORIES = Arrays.stream(SoundCategory.values()).collect(Collectors.toMap(SoundCategory::getName, Function.identity()));
+		private static final Codec<SoundCategory> SOUND_CATEGORY_CODEC = Codec.stringResolver(SoundCategory::getName, name -> CATEGORIES.getOrDefault(name.toLowerCase(Locale.ROOT), SoundCategory.NEUTRAL));
+		public static final Codec<SoundType> CODEC = RecordCodecBuilder.create(instance -> instance.group(
+				Entry.CODEC.listOf().fieldOf("sounds").forGetter(SoundType::sounds),
+				SOUND_CATEGORY_CODEC.fieldOf("category").forGetter(SoundType::category),
+				Codec.STRING.optionalFieldOf("subtitle").forGetter(SoundType::subtitle)
+		).apply(instance, SoundType::new));
+	}
+
+	private record Entry(Identifier name, RegistrationType type, float volume, float pitch, int weight, int attenuationDistance, boolean stream, boolean preload) {
+		private static final Codec<Entry> MAP_CODEC = RecordCodecBuilder.create(instance -> instance.group(
+				Identifier.CODEC.fieldOf("name").forGetter(Entry::name),
+				RegistrationType.CODEC.optionalFieldOf("type", RegistrationType.FILE).forGetter(Entry::type),
+				Codec.FLOAT.optionalFieldOf("volume", 1F).forGetter(Entry::volume),
+				Codec.FLOAT.optionalFieldOf("pitch", 1F).forGetter(Entry::pitch),
+				Codec.INT.optionalFieldOf("weight", 1).forGetter(Entry::weight),
+				Codec.INT.optionalFieldOf("attenuation_distance", 16).forGetter(Entry::attenuationDistance),
+				Codec.BOOL.optionalFieldOf("stream", false).forGetter(Entry::stream),
+				Codec.BOOL.optionalFieldOf("preload", false).forGetter(Entry::preload)
+		).apply(instance, Entry::new));
+
+		private static final Codec<Entry> STRING_CODEC = Identifier.CODEC.xmap(
+				id -> new Entry(id, RegistrationType.FILE, 1F, 1F, 1, 16, false, false),
+				Entry::name
+		);
+		private static final Codec<Entry> CODEC = Codec.xor(STRING_CODEC, MAP_CODEC).xmap(Either::unwrap, sound -> {
+			if (sound.type() != RegistrationType.FILE
+					|| sound.volume() != 1F
+					|| sound.pitch() != 1F
+					|| sound.weight() != 1
+					|| sound.attenuationDistance() != 16
+					|| sound.stream()
+					|| sound.preload()) {
+				return Either.right(sound);
+			}
+
+			return Either.left(sound);
+		});
+	}
+
+	public static final class EntryBuilderImpl implements EntryBuilder {
+		private final Identifier id;
+		private final RegistrationType type;
+
+		private float volume = 1F;
+		private float pitch = 1F;
+		private int attenuationDistance = 16;
+		private int weight = 1;
```

Comment:
Might be a good idea to create constants for these default values, as they are used in 3 places.

Reply:
Yeah that is a good idea. Also if someone else wants to use them, then they would be there.